File: Exp1_2/04_statistical_analysis/Exp2_limit_analyse.R
## Analyse the experiments of the RL emotion study
library(tidyverse)
library(lme4)
library(lmerTest)
library(Cairo)

args <- commandArgs(trailingOnly = FALSE)
file_arg <- grep("^--file=", args, value = TRUE)
script_dir <- if (length(file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", file_arg[1]), winslash = "/", mustWork = TRUE))
} else {
  getwd()
}
exp_root <- normalizePath(file.path(script_dir, ".."), winslash = "/", mustWork = TRUE)

source(file.path(script_dir, "theme-publication.R"))

# Exp2_human_limited.csv for limited human result
data <- read.table(file.path(exp_root, "data", "human_limit.csv"), header = T, sep = ";") %>%
  pivot_longer(cols = starts_with("Em."),
               names_to = "Emotion",
               names_prefix = "Em.",
               values_to = "Val")

## test the hypothesis
model <- lmer(Val ~ Story*Emotion + (1|ID), data = data)
anova(model)

## Get 95% CI (approximate, not estimated)
data.h <- data %>% 
  group_by(Story, Emotion) %>%
  summarise(sd = sd(Val), Val = mean(Val),) %>%
  mutate(SE = sd / sqrt(length(unique(data$ID))),
         CI_lower = pmax(Val - 1.96*SE, 0),
         CI_upper = pmin(Val + 1.96*SE, 1)) %>%
  dplyr::select(Story, Emotion, Val, CI_lower, CI_upper)


## Standardise emotion within stories
# data %>% group_by(Story, Emotion) %>% summarise(Val = mean(Val)) %>%
#   mutate(Val = Val/sum(Val))



# ggplot(data %>% group_by(Story, Emotion) %>% summarise(Val = mean(Val)) %>%
#          mutate(Val = Val/sum(Val)), aes(Emotion, Val)) +
#   geom_bar(stat = "identity") +
#   facet_grid(Story ~ .) +
#   ylim(0,1)


## Model SVM predictions for experiment 1.

data.m <- read.table(file.path(exp_root, "data", "svm_limit_0.014_var.csv"), header = T, sep = ",") %>%
  group_by(Story, Emotion) %>%
  summarise(Val = mean(Val)) %>%
  mutate(CI_lower = Val, CI_upper = Val)
  # summarise(Val = mean(Val))

## Visualise both data - combine model and human data
# data.b <- data %>% group_by(Story, Emotion) %>% summarise(Val = mean(Val)) %>%
#   mutate(Val = Val/sum(Val)) %>%
#   mutate(Source = "Human") %>%
#   rbind(data.m %>% mutate(Source = "Model"))

data.b <- data.h %>%
  mutate(Source = "Human") %>%
  rbind(data.m %>% mutate(Source = "Model"))


## Change the order of the x-axis and facet 
data.b$Story <- factor(data.b$Story, levels = c("Happiness","Joy","Pride",
                                        "Boredom","Fear","Sadness","Shame"))
data.b$Emotion <- factor(data.b$Emotion, levels = c("Happiness","Joy","Pride",
                                        "Boredom","Fear","Sadness","Shame"))

## Visualize both data - ggplot 

df_nonzero <- data.b[data.b$Val != 0,]


ggplot(data.b, aes(Emotion, Val, fill = Source)) +
  geom_bar(stat = "identity",width = 0.8, position = position_dodge(0.85)) +
  geom_errorbar(data = df_nonzero, aes(ymin = ifelse(Source == "Human", CI_lower, NA), 
                    ymax = ifelse(Source == "Human", CI_upper, NA)), 
                width = 0.2, position = position_dodge(0.85)) + 
  facet_grid(Story ~ .) +
  ylab("Intensity") + xlab('Emotion')+
  scale_y_continuous(limits = c(0,1),breaks = c(0,0.5,1))+
  theme_Publication() + scale_fill_Publication() + 
  theme(legend.position = c(0.92, 0.96),
        legend.margin = margin(1, 1, 1,1, "mm"),
        legend.title=element_blank(),
        legend.background = element_rect(fill='transparent'),
        panel.grid.major.x = element_blank() ,
        legend.key.size = unit(0.9, "lines"),
        legend.box.background = element_rect(colour = "black",fill="transparent"),
        legend.spacing.y = unit(0, "mm"))


ggsave(file = file.path(exp_root, "plots", "exp2_limit.jpg"), width=8, height=8,dpi=300)

ggplot(data.b, aes(Emotion, Val, fill = Source)) +
  geom_point()

## Both data, long form
data.bl <- data %>% group_by(Story, Emotion) %>% summarise(Val = mean(Val)) %>%
    mutate(Val = Val/sum(Val)) %>%
    left_join(data.m, by = c("Story","Emotion")) %>%
    mutate(d = (Val.x - Val.y)^2)

## Investigate fit
ggplot(data.bl %>% group_by(C) %>% summarise(d = mean(d)) %>%
       mutate(d = sqrt(d)),
       aes(C, d)) +
    geom_point() +
    geom_smooth()

## Scatter plot for fit
ggplot(data.bl, aes(Val.y, Val.x)) +
    geom_point() +
    geom_smooth(method = "lm") +
    ylim(0,0.5) +
    xlim(0,0.5)


summary(lm(Val.x ~ Val.y, data = data.bl))

sqrt(data.bl%>%ungroup() %>%summarise(d=mean(d)))

File: Exp1_2/04_statistical_analysis/Exp2_limit_analyse.py
import os
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import statsmodels.formula.api as smf

# ---------------------------
# Paths
# ---------------------------
script_dir = os.path.dirname(os.path.abspath(__file__))
exp_root = os.path.abspath(os.path.join(script_dir, ".."))

# ---------------------------
# Load Human Data
# ---------------------------
data = pd.read_csv(os.path.join(exp_root, "data", "human_limit.csv"), sep=";")

# Pivot (wide → long)
data = data.melt(
    id_vars=["ID", "Story"],
    value_vars=[col for col in data.columns if col.startswith("Em.")],
    var_name="Emotion",
    value_name="Val"
)

data["Emotion"] = data["Emotion"].str.replace("Em.", "")

# ---------------------------
# Mixed Effects Model
# ---------------------------
model = smf.mixedlm("Val ~ Story * Emotion", data, groups=data["ID"])
result = model.fit()
print(result.summary())

# ---------------------------
# Human Aggregation + CI
# ---------------------------
grouped = data.groupby(["Story", "Emotion"])

data_h = grouped["Val"].agg(['mean', 'std']).reset_index()
data_h.rename(columns={"mean": "Val", "std": "sd"}, inplace=True)

n = data["ID"].nunique()

data_h["SE"] = data_h["sd"] / np.sqrt(n)
data_h["CI_lower"] = np.maximum(data_h["Val"] - 1.96 * data_h["SE"], 0)
data_h["CI_upper"] = np.minimum(data_h["Val"] + 1.96 * data_h["SE"], 1)
data_h["Source"] = "Human"

# ---------------------------
# Model Data
# ---------------------------
data_m = pd.read_csv(os.path.join(exp_root, "data", "svm_limit_0.014_var.csv"))

data_m = data_m.groupby(["Story", "Emotion"])["Val"].mean().reset_index()
data_m["CI_lower"] = data_m["Val"]
data_m["CI_upper"] = data_m["Val"]
data_m["Source"] = "Model"

# ---------------------------
# Combine Data
# ---------------------------
data_b = pd.concat([data_h, data_m])

# Order categories
order = ["Happiness","Joy","Pride","Boredom","Fear","Sadness","Shame"]

data_b["Story"] = pd.Categorical(data_b["Story"], categories=order, ordered=True)
data_b["Emotion"] = pd.Categorical(data_b["Emotion"], categories=order, ordered=True)

# ---------------------------
# Remove zero values (like df_nonzero in R)
# ---------------------------
df_nonzero = data_b[data_b["Val"] != 0]

# ---------------------------
# Plot (Bar + Error bars)
# ---------------------------
g = sns.catplot(
    data=data_b,
    x="Emotion",
    y="Val",
    hue="Source",
    col="Story",
    kind="bar",
    col_wrap=1,
    height=3,
    aspect=2
)

# Add error bars only for human + non-zero
for ax, story in zip(g.axes.flatten(), order):
    subset = df_nonzero[(df_nonzero["Story"] == story) & (df_nonzero["Source"] == "Human")]

    if len(subset) > 0:
        ax.errorbar(
            x=np.arange(len(subset)),
            y=subset["Val"],
            yerr=[subset["Val"] - subset["CI_lower"], subset["CI_upper"] - subset["Val"]],
            fmt='none',
            capsize=3
        )

plt.ylim(0, 1)
plt.tight_layout()
plt.savefig(os.path.join(exp_root, "plots", "exp2_limit.png"), dpi=300)

# ---------------------------
# Long Form Comparison
# ---------------------------
data_bl = data.groupby(["Story", "Emotion"])["Val"].mean().reset_index()
data_bl["Val"] = data_bl.groupby("Story")["Val"].transform(lambda x: x / x.sum())

data_bl = data_bl.merge(data_m, on=["Story", "Emotion"], suffixes=("_x", "_y"))
data_bl["d"] = (data_bl["Val_x"] - data_bl["Val_y"])**2

# RMSE
rmse = np.sqrt(data_bl["d"].mean())
print("RMSE:", rmse)

# ---------------------------
# Scatter Plot
# ---------------------------
sns.regplot(data=data_bl, x="Val_y", y="Val_x")
plt.xlim(0, 0.5)
plt.ylim(0, 0.5)
plt.savefig(os.path.join(exp_root, "plots", "exp2_scatter.png"))

# ---------------------------
# Linear Regression Summary
# ---------------------------
lm_model = smf.ols("Val_x ~ Val_y", data=data_bl).fit()
print(lm_model.summary())